### PR TITLE
New plastic behavior: shear slip on joint/plane of weakness

### PR DESCRIPTION
**Pull request by lurpi**
Documentation will follow soon, including analytical validation. 
Modification to allows calculation of shear failure along a predefined direction
Input file .msp looks like this:

$PLASTICITY                     
   MOHR-COULOMB 
   15.e6         //cohesion
   40           //friction angle
   0.0          //dilation angle        
   3.5e16        //tensile strength
   0            //curve number for strain hardening of cohesion
   0            //curve number for strain hardening of friction angle  
$WEAKNESS_PLANE
  MICRO_STRUCTURE_TENSOR 1 1 1  //no microstruct
  WEAKPLANE_NORM   0 -0.984807753 0.17364818 // plane of weakness normal
   1.45e6         //joint cohesion
   28           //joint friction angle
   0.0          //joint dilation angle         
   0.3.e6        //joint tensile strength
   0            //curve number for strain hardening of joint cohesion
   0            //curve number for strain hardening of joint friction angle

### CODE REVIEW

**Comment by lurpi** [comment]
rf_msp.cpp modification are not correctly noted.
My modifications are

- [(from line 1042](https://github.com/lurpi/ogs5/blob/c5d15d5215bf18d94d3b77c5305c6b01a9bd41b5/FEM/rf_msp_new.cpp#L1042) to line [1086, setting parameters of the joint](https://github.com/lurpi/ogs5/blob/c5d15d5215bf18d94d3b77c5305c6b01a9bd41b5/FEM/rf_msp_new.cpp#L1086)

- [Function frictional coefficent of the joint CSolidProperties::CalculateCoefficent_MOHRjoint](https://github.com/lurpi/ogs5/blob/c5d15d5215bf18d94d3b77c5305c6b01a9bd41b5/FEM/rf_msp_new.cpp#L2812)

- [function StressIntegrationMOHR_Joint](https://github.com/lurpi/ogs5/blob/c5d15d5215bf18d94d3b77c5305c6b01a9bd41b5/FEM/rf_msp_new.cpp#L3760) to evaluate mohr coulomb rupture in the matrix and sequentially in the joint

**Comment by wenqing** [comment]
There are compilation errors. See: https://travis-ci.org/ufz/ogs5/jobs/608123326?utm_medium=notification&utm_source=github_status

File: FEM/rf_msp_new.cpp [hunk 1]
```diff
@@ -4239,7 +4251,14 @@ int CSolidProperties::StressIntegrationMOHR_Joint(const int GPiGPj, const Elemen
         *TmpMatrix = (0.);
         *TmpMatrix2 = (0.);
         CalculateCoefficent_MOHRjoint(ep, 0, 0);
-        tensionsurf = -1; //TryStr_buff[2]; // tensile behavior not yet included
+        if (GPiGPj == 5)
+            {
+            std::cout << "Princ, 1 (vert) -5.50e+06 , 2 (along strike) -3.70e+06 ,  stress 3 (perp to strike) -2.80e+06  for elem " << (*ele_val->GetIndex(), ele_value_id) << endl;
+            std::cout << "OGS , stress_xx " << (*ele_val->Stress)(0, GPiGPj) << " , stress_yy " << (*ele_val->Stress)(1, GPiGPj) << " , stress_zz " << (*ele_val->Stress)(2, GPiGPj) << " , stress_xy " << (*ele_val->Stress)(3, GPiGPj) << " , stress_xz " << (*ele_val->Stress)(4, GPiGPj) << " , stress_yz " << (*ele_val->Stress)(5, GPiGPj) << endl;
+            std::cout << "stress local_11 " << TryStr_buff[0] << " , stress local_22 " << TryStr_buff[1] << " , stress local_33 " << TryStr_buff[2] << " SHEAR local_12 " << TryStr_buff[3] << " , local_13 " << TryStr_buff[4] << " , local_23  " << TryStr_buff[5] << endl;
+            shearsurf = sqrt((TryStr_buff[5] * TryStr_buff[5]) + (TryStr_buff[4] * TryStr_buff[4])) + TryStr_buff[2] * tan(thetaj) - Y0j;
```
[the remark this answers is not in the document]
Comment (lurpi) [reply FEM/rf_msp_new.cpp]: Dummy error. Correcting the source file

File: FEM/rf_msp_new.cpp [hunk 2]
```diff
@@ -4250,10 +4269,28 @@ int CSolidProperties::StressIntegrationMOHR_Joint(const int GPiGPj, const Elemen
             //id_fs_s[4] =-1 ;
             }
         double shearstress = sqrt((TryStr_buff[5] * TryStr_buff[5]) + (TryStr_buff[4] * TryStr_buff[4]));
-        
+        if (fabs(shearstress < 150000 ) || ep > 0.0035)
```
[the remark this answers is not in the document]
Comment (lurpi) [reply FEM/rf_msp_new.cpp]: specific values for troublesome simulation. Not necessary, will be removed in the corrected file

**Comment by lurpi** [comment]
> clang-format has to be applied to the source code. If you use visual studio 2017/2019, you can download clang-format extension:
> https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.ClangFormat

I used a standalone build, available at the following url 
http://llvm.org/builds/
and the .clang-format provided in the ogs5 ufz master branch, but there are still some differences (the header file is completely not recognized?)

**Comment by wenqing** [comment]
if you can run command line of clang-format, you can try

clang-format -i -sort-includes=false -style=file [path]/rf_msp_new.h
clang-format -i -sort-includes=false -style=file [path]/rf_msp_new.cpp

**Comment by lurpi** [comment]
thanks Wenqing, that solved most of the issues with the header.

File: FEM/rf_msp_new.cpp [hunk 3]
```diff
@@ -1239,6 +1295,7 @@ CSolidProperties::~CSolidProperties()
     data_Density = NULL;
     data_Youngs = NULL;
     data_Plasticity = NULL;
+    data_Plasticity_joint = NULL;
```
Comment (wenqing) [inline FEM/rf_msp_new.cpp]: Please copy this line to the constructor as well, i.e before line 1239. The uninitialized pointer causes problem in the destructor under MPI and PETSc.

File: FEM/rf_msp_new.cpp [hunk 4]
```diff
@@ -3684,6 +3758,805 @@ void CSolidProperties::TangentialDPwithTensionCorner(Matrix* Dep, double /*mm*/)
...
+                                          //
+                for (i = 0; i < Size; i++)
+                {
+                    for (j = 0; j < Size; j++)
+                        (*TmpDe)(i, j) -=
+                            (*TmpMatrix2)(i, j) /
+                            TmpValue1;  // Dep = De - De dG/ds dF/ds De / dF/ds
+                                        // de dG/ds
+                }
+                // TEST
```
Comment (wenqing) [inline FEM/rf_msp_new.cpp]: Unused comments.

File: FEM/rf_msp_new.cpp [hunk 5]
```diff
@@ -3684,6 +3758,805 @@ void CSolidProperties::TangentialDPwithTensionCorner(Matrix* Dep, double /*mm*/)
...
+            for (i = 0; i < Size; i++)
+                TryStress[i] = 0;
+            TransMicroStru_T->multi(TmpStress, TryStress);
+
+            *ConstitutiveMatrix = (0.);
+            TransMicroStru_T->multi(*TmpDe, *TransMicroStru,
+                                    *ConstitutiveMatrix);
+            // TEST
+            // ConstitutiveMatrix->Write();
+            //
```
Comment (wenqing) [inline FEM/rf_msp_new.cpp]: Unused comments.

**Comment by wenqing** [comment]
@norihiro-w Do you have any comment? If not, I  will merge it.

**Comment by norihiro-w** [comment]
@wenqing i have no further comments.

**Pull request merged**